File: scala/sparker/src/main/scala-2.11/SparkER/BlockRefinementMethods/PruningMethods/CEP.scala
package SparkER.BlockRefinementMethods.PruningMethods

import SparkER.BlockRefinementMethods.PruningMethods.PruningUtils.WeightTypes
import SparkER.DataStructures.{ProfileBlocks, UnweightedEdge}

import scala.collection.mutable
//import org.apache.log4j.LogManager
import org.apache.spark.SparkContext
import org.apache.spark.broadcast.Broadcast
import org.apache.spark.rdd.RDD

/**
  * Created by Luca on 10/04/2018.
  */
object CEP {
  /**
    * Computes the Cardinality Edge Pruning
    *
    * @param profileBlocksFiltered  profileBlocks after block filtering
    * @param blockIndex             a map that given a block ID returns the ID of the contained profiles
    * @param maxID                  maximum profile ID
    * @param separatorIDs           maximum profile ID of the first dataset (-1 if there is only one dataset)
    * @param groundtruth            set of true matches
    * @param weightType             type of weight to use @see BlockRefinementMethods.PruningMethods.PruningUtils.WeightTypes
    * @param profileBlocksSizeIndex a map that contains for each profile the number of its blocks
    * @param useEntropy             if true use the provided entropies to improve the edge weighting
    * @param blocksEntropies        a map that contains for each block its entropy
    * @return an RDD that contains for each partition the number of existing edges, number of gt edges, and the retained edges
    **/
  def CEP(profileBlocksFiltered: RDD[ProfileBlocks],
          blockIndex: Broadcast[scala.collection.Map[Long, Array[Set[Long]]]],
          maxID: Int,
          separatorIDs: Array[Long],
          groundtruth: Broadcast[scala.collection.immutable.Set[(Long, Long)]],
          weightType: String = WeightTypes.CBS,
          profileBlocksSizeIndex: Broadcast[scala.collection.Map[Long, Int]] = null,
          useEntropy: Boolean = false,
          blocksEntropies: Broadcast[scala.collection.Map[Long, Double]] = null
         )
  : RDD[(Double, Double, Iterable[UnweightedEdge])] = {

    val edgesToKeep = math.floor(blockIndex.value.values.flatMap(_.map(_.size.toDouble)).sum / 2.0)

    println("SPARKER - Number of edges to keep " + edgesToKeep)

    if (useEntropy && blocksEntropies == null) {
      throw new Exception("blocksEntropies must be defined")
    }

    if ((weightType == WeightTypes.ECBS || weightType == WeightTypes.EJS || weightType == WeightTypes.JS || weightType == WeightTypes.chiSquare) && profileBlocksSizeIndex == null) {
      throw new Exception("profileBlocksSizeIndex must be defined")
    }

    if (!List(WeightTypes.CBS, WeightTypes.JS, WeightTypes.chiSquare, WeightTypes.ARCS, WeightTypes.ECBS, WeightTypes.EJS).contains(weightType)) {
      throw new Exception("Please provide a valid WeightType, " + weightType + " is not an acceptable value!")
    }

    val sc = SparkContext.getOrCreate()
    var numberOfEdges: Double = 0
    var edgesPerProfile: Broadcast[scala.collection.Map[Long, Double]] = null

    if (weightType == WeightTypes.EJS) {
      val stats = CommonNodePruning.computeStatistics(profileBlocksFiltered, blockIndex, maxID, separatorIDs)
      numberOfEdges = stats.map(_._1.toDouble).sum()
      edgesPerProfile = sc.broadcast(stats.map(x => (x._2._1, x._2._2.toDouble)).groupByKey().map(x => (x._1, x._2.sum)).collectAsMap())
    }

    val threshold = calcThreshold(profileBlocksFiltered, blockIndex, maxID, separatorIDs, weightType, profileBlocksSizeIndex, useEntropy, blocksEntropies, numberOfEdges, edgesPerProfile, edgesToKeep)
    println("SPARKER - Computed threshold " + threshold)


    val toKeep = sc.broadcast(threshold._2)

    val edges = pruning(profileBlocksFiltered, blockIndex, maxID, separatorIDs, groundtruth, weightType, profileBlocksSizeIndex, useEntropy, blocksEntropies, threshold._1, toKeep, numberOfEdges, edgesPerProfile)


    toKeep.unpersist()
    if (edgesPerProfile != null) {
      edgesPerProfile.unpersist()
    }

    edges
  }


  /**
    * Compute the threshold for a profile
    *
    * @param weights         an array which contains the weight of each neighbor
    * @param neighbors       an array which contains the IDs of the neighbors
    * @param neighborsNumber number of neighbors
    * @return the frequencies for each weight
    **/
  def calcFreq(weights: Array[Double],
               neighbors: Array[Int],
               neighborsNumber: Int): Array[(Double, Double)] = {
    var acc: Double = 0
    val out = mutable.Map[Double, Double]()
    for (i <- 0 until neighborsNumber) {
      val weight = weights(neighbors(i))
      val w = out.getOrElse(weight, 0.0) + 1
      out.put(weight, w)
    }
    out.toArray
  }

  /**
    * Performs the pruning
    *
    * @param weights         an array which contains the weight of each neighbor
    * @param neighbors       an array which  contains the IDs of the neighbors
    * @param neighborsNumber number of neighbors
    * @param groundtruth     set of true matches
    * @param weightType      type of weight to use @see BlockRefinementMethods.PruningMethods.PruningUtils.WeightTypes
    * @param globalThreshold global threshold
    * @return a tuple that contains the number of retained edges and the edges that exists in the groundtruth
    **/
  def doPruning(profileID: Long,
                weights: Array[Double],
                neighbors: Array[Int],
                neighborsNumber: Int,
                groundtruth: Broadcast[scala.collection.immutable.Set[(Long, Long)]],
                weightType: String,
                globalThreshold: Double,
                toKeep: Broadcast[scala.collection.Map[Long, Double]]
               ): (Double, Double, Iterable[UnweightedEdge]) = {

    var cont: Double = 0
    var gtFound: Double = 0
    var edges: List[UnweightedEdge] = Nil

    var neighborToKeep = toKeep.value.getOrElse(profileID, 0.0).toInt

    //@transient lazy val log = LogManager.getRootLogger

    for (i <- 0 until neighborsNumber) {
      val neighborID = neighbors(i)
      val neighborWeight = weights(neighborID)

      if ((neighborWeight > globalThreshold) || (neighborWeight == globalThreshold && neighborToKeep > 0)) {
        //log.info("SPARKER - Id mio "+profileID+", ID vicino "+neighborID+", soglia vicino "+neighborThreshold+", peso vicino "+neighborWeight+", soglia mia "+profileThreshold+", comparison type "+comparisonType+" ----> lo tengo")
        cont += 1
        if (groundtruth != null && groundtruth.value.contains((profileID, neighborID))) {
          gtFound += 1
        }

        edges = UnweightedEdge(profileID, neighbors(i)) :: edges

        if (neighborWeight == globalThreshold) {
          neighborToKeep -= 1
        }
      } /*
        else{
          log.info("SPARKER - Id mio "+profileID+", ID vicino "+neighborID+", soglia vicino "+neighborThreshold+", peso vicino "+neighborWeight+", soglia mia "+profileThreshold+", comparison type "+comparisonType+" ----> Non lo tengo")
        }*/
    }

    (cont, gtFound, edges)
  }


  /**
    * Computes the global threshold
    *
    * @param profileBlocksFiltered  profileBlocks after block filtering
    * @param blockIndex             a map that given a block ID returns the ID of the contained profiles
    * @param maxID                  maximum profile ID
    * @param separatorID            maximum profile ID of the first dataset (-1 if there is only one dataset)
    * @param weightType             type of weight to use @see BlockRefinementMethods.PruningMethods.PruningUtils.WeightTypes
    * @param profileBlocksSizeIndex a map that contains for each profile the number of its blocks
    * @param useEntropy             if true use the provided entropies to improve the edge weighting
    * @param blocksEntropies        a map that contains for each block its entropy
    * @param numberOfEdges          global number of existings edges
    * @param edgesPerProfile        a map that contains for each profile the number of edges
    * @return the global threshold that will be used to prune the graph
    **/
  def calcThreshold(profileBlocksFiltered: RDD[ProfileBlocks],
                    blockIndex: Broadcast[scala.collection.Map[Long, Array[Set[Long]]]],
                    maxID: Int,
                    separatorID: Array[Long],
                    weightType: String,
                    profileBlocksSizeIndex: Broadcast[scala.collection.Map[Long, Int]],
                    useEntropy: Boolean,
                    blocksEntropies: Broadcast[scala.collection.Map[Long, Double]],
                    numberOfEdges: Double,
                    edgesPerProfile: Broadcast[scala.collection.Map[Long, Double]],
                    edgesToKeep: Double
                   ): (Double, scala.collection.Map[Long, Double]) = {

    val partialFreqs = profileBlocksFiltered.mapPartitions { partition =>
      val localWeights = Array.fill[Double](maxID + 1) {
        0
      }
      val neighbors = Array.ofDim[Int](maxID + 1)
      var neighborsNumber = 0

      val entropies: Array[Double] = {
        if (useEntropy) {
          Array.fill[Double](maxID + 1) {
            0.0
          }
        }
        else {
          null
        }
      }

      partition.map { pb =>
        neighborsNumber = CommonNodePruning.calcCBS(pb, blockIndex, separatorID, useEntropy, blocksEntropies, localWeights, entropies, neighbors, false)
        CommonNodePruning.calcWeights(pb, localWeights, neighbors, entropies, neighborsNumber, blockIndex, separatorID, weightType, profileBlocksSizeIndex, useEntropy, numberOfEdges, edgesPerProfile)
        val localFreq = calcFreq(localWeights, neighbors, neighborsNumber)
        CommonNodePruning.doReset(localWeights, neighbors, entropies, useEntropy, neighborsNumber)
        (pb.profileID, localFreq.toList)
      }
    }

    val frequencies = partialFreqs.flatMap(x => x._2).groupByKey().map(x => (x._1, x._2.sum)).collect().sortBy(-_._1)

    var sum = 0.0
    var cont = 0
    while (sum < edgesToKeep && cont < frequencies.length) {
      sum += frequencies(cont)._2
      cont += 1
    }

    val threshold = frequencies(cont - 1)._1
    var lastToKeep = frequencies(cont - 1)._2 - (sum - edgesToKeep)

    println("SPARKER - Di quelli con peso uguale a  " + threshold + " devo tenerne " + lastToKeep)


    /** Per ogni profilo conto quanti elementi ha dell'ultima soglia che devo tenere */
    val stats = partialFreqs.map { p =>
      (p._1, p._2.filter(_._1 == threshold).map(_._2))
    }.filter(_._2.nonEmpty).map(x => (x._1, x._2.head)).collect()


    /** Scelgo a caso per ogni profilo quanti tenerne in modo da tenere esattamente il numero che mi serve */
    val toKeep = for (i <- stats.indices; if lastToKeep > 0) yield {
      if (stats(i)._2 > lastToKeep) {
        val a = lastToKeep
        lastToKeep -= stats(i)._2
        (stats(i)._1, a)
      }
      else {
        lastToKeep -= stats(i)._2
        stats(i)
      }
    }

    (threshold, toKeep.filter(_._2 > 0).toMap)
  }

  /**
    * Performs the pruning
    *
    * @param profileBlocksFiltered  profileBlocks after block filtering
    * @param blockIndex             a map that given a block ID returns the ID of the contained profiles
    * @param maxID                  maximum profile ID
    * @param separatorID            maximum profile ID of the first dataset (-1 if there is only one dataset)
    * @param groundtruth            set of true matches
    * @param weightType             type of weight to use @see BlockRefinementMethods.PruningMethods.PruningUtils.WeightTypes
    * @param profileBlocksSizeIndex a map that contains for each profile the number of its blocks
    * @param useEntropy             if true use the provided entropies to improve the edge weighting
    * @param blocksEntropies        a map that contains for each block its entropy
    * @param numberOfEdges          global number of existings edges
    * @param edgesPerProfile        a map that contains for each profile the number of edges
    * @return an RDD that for each partition contains the number of retained edges and the list of the elements that appears in the groundtruth
    **/
  def pruning(profileBlocksFiltered: RDD[ProfileBlocks],
              blockIndex: Broadcast[scala.collection.Map[Long, Array[Set[Long]]]],
              maxID: Int,
              separatorID: Array[Long],
              groundtruth: Broadcast[scala.collection.immutable.Set[(Long, Long)]],
              weightType: String,
              profileBlocksSizeIndex: Broadcast[scala.collection.Map[Long, Int]],
              useEntropy: Boolean,
              blocksEntropies: Broadcast[scala.collection.Map[Long, Double]],
              threshold: Double,
              toKeep: Broadcast[scala.collection.Map[Long, Double]],
              numberOfEdges: Double,
              edgesPerProfile: Broadcast[scala.collection.Map[Long, Double]]
             ): RDD[(Double, Double, Iterable[UnweightedEdge])] = {

    profileBlocksFiltered.mapPartitions { partition =>
      val localWeights = Array.fill[Double](maxID + 1) {
        0
      }
      val neighbors = Array.ofDim[Int](maxID + 1)
      var neighborsNumber = 0

      val entropies: Array[Double] = {
        if (useEntropy) {
          Array.fill[Double](maxID + 1) {
            0.0
          }
        }
        else {
          null
        }
      }

      partition.map { pb =>
        neighborsNumber = CommonNodePruning.calcCBS(pb, blockIndex, separatorID, useEntropy, blocksEntropies, localWeights, entropies, neighbors, false)
        CommonNodePruning.calcWeights(pb, localWeights, neighbors, entropies, neighborsNumber, blockIndex, separatorID, weightType, profileBlocksSizeIndex, useEntropy, numberOfEdges, edgesPerProfile)
        val result = doPruning(pb.profileID, localWeights, neighbors, neighborsNumber, groundtruth, weightType, threshold, toKeep)
        CommonNodePruning.doReset(localWeights, neighbors, entropies, useEntropy, neighborsNumber)
        result
      }
    }
  }
}

File: python/sparker/cep.py
from .pruning_utils import WeightTypes, ThresholdTypes, ComparisonTypes
from .common_node_pruning import compute_statistics, calc_cbs, calc_weights, do_reset
import math


class CEP(object):

    @staticmethod
    def do_pruning(profile_id, weights, neighbors, neighbors_number, groundtruth, threshold, edges_to_keep):
        """
        Performs the pruning
        :param profile_id: id of the profile
        :param weights: an array which contains the weight of each neighbor
        :param neighbors: an array which  contains the IDs of the neighbors
        :param neighbors_number: number of neighbors
        :param groundtruth: set of true matches (optional)
        :param threshold: global threshold
        :param edges_to_keep: number of edges to keep with a weight equal to global threshold for every profile
        :return: a triplet that contains the number of retained edges, the edges that exists in the groundtruth,
                 and the retained edges
        """
        num_edges = 0
        gt_found = 0
        edges = []
        neighbors_to_keep = 0
        if profile_id in edges_to_keep.value:
            neighbors_to_keep = edges_to_keep.value[profile_id]

        for i in range(0, neighbors_number):
            neighbor_id = neighbors[i]
            neighbor_weight = weights[neighbor_id]
            if (neighbor_weight > threshold) or (neighbor_weight == threshold and neighbors_to_keep > 0):
                num_edges += 1
                if groundtruth is not None:
                    if (profile_id, neighbor_id) in groundtruth.value:
                        gt_found += 1
                edges.append((profile_id, neighbor_id, neighbor_weight))

                if neighbor_weight == threshold:
                    neighbors_to_keep -= 1

        return num_edges, gt_found, edges

    @staticmethod
    def pruning(profile_blocks_filtered, block_index, max_id, separator_id, groundtruth, weight_type,
                profile_blocks_size_index, use_entropy, blocks_entropies, threshold, edges_to_keep, number_of_edges,
                edges_per_profile):
        """
        Performs the pruning
        :param profile_blocks_filtered: profile_blocks after block filtering
        :param block_index: a map that given a block ID returns the ID of the contained profiles
        :param max_id: maximum profile ID
        :param separator_id: maximum profile ID of each dataset (-1 if there is only one dataset)
        :param groundtruth: set of true matches
        :param weight_type: type of weight to use see pruning_utils.WeightTypes
        :param profile_blocks_size_index: a map that contains for each profile the number of its blocks
        :param use_entropy: if true use the provided entropies to improve the edge weighting
        :param blocks_entropies: a map that contains for each block its entropy
        :param threshold: global threshold
        :param edges_to_keep: number of edges to keep with a weight equal to global threshold for every profile
        :param number_of_edges: global number of existing edges
        :param edges_per_profile: a map that contains for each profile the number of edges
        :return: an RDD that for each partition contains a triplet that contains the number of retained edges,
        the edges that exists in the groundtruth, and the retained edges
        """

        def compute_partition(partition):
            local_weights = [0] * (max_id+1)
            neighbors = [0] * (max_id+1)
            entropies = None
            if use_entropy:
                entropies = [0.0] * (max_id+1)

            def inside_map(pb):
                neighbors_number = calc_cbs(pb, block_index, separator_id, use_entropy, blocks_entropies,
                                            local_weights, entropies, neighbors, False)
                calc_weights(pb, local_weights, neighbors, entropies, neighbors_number, block_index, separator_id,
                             weight_type, profile_blocks_size_index, use_entropy, number_of_edges, edges_per_profile)
                result = CEP.do_pruning(pb.profile_id, local_weights, neighbors, neighbors_number, groundtruth,
                                        threshold, edges_to_keep)
                do_reset(local_weights, neighbors, entropies, use_entropy, neighbors_number)
                return result

            return map(inside_map, partition)

        return profile_blocks_filtered.mapPartitions(compute_partition)

    @staticmethod
    def calc_threshold(weights, neighbors, neighbors_number, threshold_type):
        """
        Computes the threshold for a profile
        :param weights: an array which contains the weight of each neighbor
        :param neighbors: an array which contains the IDs of the neighbors
        :param neighbors_number: number of neighbors
        :param threshold_type: type of threshold to use
        :return: the profile' threshold
        """
        acc = 0
        for i in range(0, neighbors_number):
            neighbor_id = neighbors[i]
            if threshold_type == ThresholdTypes.AVG:
                acc += weights[neighbor_id]
            elif threshold_type == ThresholdTypes.MAX_FRACT_2 and weights[neighbor_id] > acc:
                acc = weights[neighbor_id]
            else:
                pass

        if threshold_type == ThresholdTypes.AVG:
            if neighbors_number > 0:
                acc /= float(neighbors_number)
        else:
            acc /= 2.0

        return acc

    @staticmethod
    def calc_weights_frequencies(weights, neighbors, neighbors_number):
        """
        Computes the number of neighbors for every weight.
        :param weights: list of weights for every neighbor
        :param neighbors: list of neighbors
        :param neighbors_number: number of neighbors
        :return: number of neighbors for every weight
        """
        out = {}
        for i in range(0, neighbors_number):
            weight = weights[neighbors[i]]
            if weight in out:
                out[weight] += 1
            else:
                out[weight] = 1
        return list(out.items())

    @staticmethod
    def calc_thresholds(profile_blocks_filtered,
                        block_index,
                        max_id,
                        separator_ids,
                        weight_type,
                        profile_blocks_size_index,
                        use_entropy,
                        blocks_entropies,
                        number_of_edges,
                        edges_per_profile,
                        num_of_edges_to_keep
                        ):
        """
         Given the number of edges to keep, for each profile calculate the weights of the neighbors.
         For each weight, its global frequency is calculated, i.e. for every weight how many edges are associated.
         Then, starting from the highest weight, it compute the weight that let to reach a sufficient number of edges,
         i.e. the weight that adding the edges of that weight would exceed the maximum number of edges to keep.
         For that threshold, for each profile it calculates how many edges must be kept.

         What it returns is:
           - The weight for which I have to hold all the edges> of that weight
           - A list showing for each profile how many edges it must keep with that weight

        :param profile_blocks_filtered:
        :param block_index:
        :param max_id:
        :param separator_ids:
        :param weight_type:
        :param profile_blocks_size_index:
        :param use_entropy:
        :param blocks_entropies:
        :param number_of_edges:
        :param edges_per_profile:
        :param num_of_edges_to_keep:
        :return:
        """

        def compute_partition(part):
            local_weights = [0.0] * (max_id + 1)
            neighbors = [0] * (max_id + 1)
            entropies = None
            if use_entropy:
                entropies = [0.0] * (max_id + 1)

            def inside_map(pb):
                neighbors_number = calc_cbs(pb, block_index, separator_ids, use_entropy, blocks_entropies,
                                            local_weights, entropies, neighbors, False)
                calc_weights(pb, local_weights, neighbors, entropies, neighbors_number, block_index, separator_ids,
                             weight_type, profile_blocks_size_index, use_entropy, number_of_edges, edges_per_profile)
                local_freq = CEP.calc_weights_frequencies(local_weights, neighbors, neighbors_number)
                do_reset(local_weights, neighbors, entropies, use_entropy, neighbors_number)
                return pb.profile_id, local_freq

            return map(inside_map, part)

        partial_freq = profile_blocks_filtered.mapPartitions(compute_partition)

        # Sorted list of tuples that contain for every edge weight the number of associated edges
        frequencies = sorted(
            partial_freq.flatMap(lambda x: x[1]).groupByKey().map(lambda x: (x[0], sum(x[1]))).collect(),
            key=lambda x: -x[0])

        sums = 0
        cont = 0
        # Continue while the sum of frequencies is lower than the number of edges to keep
        while sums < num_of_edges_to_keep and cont < len(frequencies):
            sums += frequencies[cont][1]
            cont += 1

        # All the edges with a weight greater than this threshold must be kept
        threshold = frequencies[cont - 1][0]
        # Number of edges to keep in the last level
        remaining_edges_to_keep = frequencies[cont - 1][1] - (sums - num_of_edges_to_keep)

        # For every profile, counts the number of edges with a weight equal to the threshold
        stats = partial_freq.map(lambda pr: (pr[0], list(map(lambda x: x[1],
                                                             filter(lambda x: x[0] == threshold, pr[1])))))\
            .filter(lambda x: len(x[1]) > 0).map(lambda x: (x[0], x[1][0])).collect()

        # For every profile, reports the number of edges to keep with a weight equal to the threshold
        num_edges_to_keep = []
        for p in stats:
            # Stop when there are no more edges to keep
            if remaining_edges_to_keep <= 0:
                break

            if p[1] > remaining_edges_to_keep:
                num_edges_to_keep.append((p[0], remaining_edges_to_keep))
            else:
                num_edges_to_keep.append(p)

            remaining_edges_to_keep -= p[1]

        return threshold, dict(num_edges_to_keep)

    @staticmethod
    def cep(profile_blocks_filtered, block_index, max_id, separator_ids=None, groundtruth=None,
            weight_type=WeightTypes.CBS, profile_blocks_size_index=None,
            use_entropy=False, blocks_entropies=None, comparison_type=ComparisonTypes.OR):
        """
        Performs the Weight Node Pruning. Returns an RDD that contains for each partition (number of edges after the
        pruning, number of true matches found in the groundtruth, list of edges)
        :param profile_blocks_filtered: profiles after filtering
        :param block_index: broadcasted blocking index
        :param max_id: highest profile ID
        :param separator_ids: id of the separators that identifies the different data sources
        :param groundtruth: groundtruth (optional)
        :param weight_type: weight method from WeightTypes, default CBS
        :param profile_blocks_size_index: broadcast map that given a block ID returns its size, needed for ECBS, EJS,
               JS and chi_square weights. (optional)
        :param use_entropy: boolean, use entropy to weight the edges. Default false.
        :param blocks_entropies: broadcasted entropies, a map that given a block return its entropy. (optional)
        :param comparison_type: pruning strategy from ComparisonTypes, default OR
        :return: Returns an RDD that contains for each partition (number of edges after the pruning, number of true
                 matches found in the groundtruth, list of edges)
        """
        if separator_ids is None:
            separator_ids = []
        valid_weights = [WeightTypes.CBS, WeightTypes.JS, WeightTypes.CHI_SQUARE, WeightTypes.ARCS,
                         WeightTypes.ECBS, WeightTypes.EJS]
        valid_comp_types = [ComparisonTypes.OR, ComparisonTypes.AND]

        if weight_type not in valid_weights:
            raise ValueError("Please provide a valid Weight_type, " + str(weight_type) + " is not an acceptable value!")

        if comparison_type not in valid_comp_types:
            raise ValueError(
                "Please provide a valid Comparison_type, " + str(comparison_type) + " is not an acceptable value!")

        if use_entropy and blocks_entropies is None:
            raise ValueError('blocks_entropies must be defined')

        if (weight_type == WeightTypes.ECBS or weight_type == WeightTypes.EJS or weight_type == WeightTypes.JS or
                weight_type == WeightTypes.CHI_SQUARE) and profile_blocks_size_index is None:
            raise ValueError('profile_blocks_size_index must be defined')

        sc = profile_blocks_filtered.context
        number_of_edges = 0
        edges_per_profile = None

        if weight_type == WeightTypes.EJS:
            stats = compute_statistics(profile_blocks_filtered, block_index, max_id, separator_ids)
            number_of_edges = stats.map(lambda x: x[0]).sum()
            edges_per_profile = sc.broadcast(
                stats.map(lambda x: (x[1][0], x[1][1])).groupByKey().map(lambda x: (x[0], sum(x[1])))
                .collectAsMap())

        num_of_edges_to_keep = math.floor(sum([sum(map(lambda y: len(y), block_index.value[x]))
                                          for x in block_index.value]) / 2.0)

        threshold, num_edges_to_keep = CEP.calc_thresholds(profile_blocks_filtered, block_index, max_id, separator_ids,
                                                           weight_type, profile_blocks_size_index, use_entropy,
                                                           blocks_entropies, number_of_edges, edges_per_profile,
                                                           num_of_edges_to_keep)

        edges_to_keep = sc.broadcast(num_edges_to_keep)

        edges = CEP.pruning(profile_blocks_filtered, block_index, max_id, separator_ids, groundtruth, weight_type,
                            profile_blocks_size_index, use_entropy, blocks_entropies, threshold, edges_to_keep,
                            number_of_edges, edges_per_profile)

        edges_to_keep.unpersist()
        if edges_per_profile is not None:
            edges_per_profile.unpersist()

        return edges
